### PR TITLE
X11 migration

### PR DESCRIPTION
**Pull request by tschoonj**
moving Formulas that depend on `gtk+` or `gtk+3` from x11 to core and other taps

see also Homebrew/homebrew-x11#176, Homebrew/homebrew#3198 and Homebrew/homebrew-gui#2

### CODE REVIEW

**Comment by mikemcquaid** [comment]
A thought that some of these may be a better fit for https://github.com/homebrew/homebrew-gui

File: Library/Formula/bochs.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  sha256 "79700ef0914a0973f62d9908ff700ef7def62d4a28ed5de418ef61f3576585ce"
+
+  bottle do
+    sha256 "9efec04a05953eeb5a385b28ed1c8a2202e13bedcfcee3933baf32cf1eadfe80" => :yosemite
+    sha256 "43d3b3d637710ea229ef456716bc3b126e32d19e6b653ec4d4e061100e79af34" => :mavericks
+    sha256 "34a83396abc3570523eb4ea9d70d33dba353097228d03de7f5879af83a15443a" => :mountain_lion
+  end
+
+  depends_on "pkg-config" => :build
+  depends_on :x11
```
Comment (DomT4) [inline Library/Formula/bochs.rb]: This formula still has a hard X11 dep.

Comment (tschoonj) [inline Library/Formula/bochs.rb]: Is this really an issue? There are several other formulas in core with a hard X11 dependency...

**Comment by tschoonj** [comment]
@mikemcquaid Is that tap still around? The link gives a 404 error and it doesn't appear in the list of Homebrew repos

File: Library/Formula/bochs.rb (hunk 1 above)
Comment (mikemcquaid) [inline Library/Formula/bochs.rb]: Yep. We should be migrating them all to homebrew/x11; that's kinda the point of the tap.

**Comment by mikemcquaid** [comment]
> The link gives a 404 error and it doesn't appear in the list of Homebrew repos

Fixed now.

File: Library/Formula/bochs.rb (hunk 1 above)
Comment (DomT4) [inline Library/Formula/bochs.rb]: Things in the core that still have a hard X11 dependency tend to be still here because other stuff relies on it.

**Comment by tschoonj** [comment]
What exactly would be the criteria for a formula to end up in homebrew-gui: some packages consist of both a library and a gui. And what to do in a case like `bochs`, which has a hard x11 dependency and is a gui :smile: ?

Either way, I assume that the vast majority of gtk- and qt- based formulas will end up being migrated to homebrew-gui?

**Comment by DomT4** [comment]
> What exactly would be the criteria for a formula to end up in homebrew-gui: some packages consist of both a library and a gui. 

This is why I made the GUI tap private for a while, to give us time to sort out what we actually wanted in there before people started randomly submitting stuff heh. I don't think we ever did, really.

My preference is more or less:
* Anything that needs X11 by default => homebrew/x11
* Anything that ships a binary that launches only a GUI, or only an `.app` bundle => homebrew/gui
* Anything else => core

**Comment by tschoonj** [comment]
What is the rationale behind a separate gui tap? I can easily understand why a science tap is useful since only a subset of Homebrew users will ever need that. The gui tap on the other hand will be used by most Homebrew users I think.

**Comment by mikemcquaid** [comment]
> The gui tap on the other hand will be used by most Homebrew users I think.

I don't think so. I (and a most programmers I've met) use Homebrew only for command-line utilities, development dependencies and/or libraries. `brew cask` is a much better fit for GUI software and, as far as I'm concerned, stuff that doesn't ship an self-contained `.app` binary for OS X isn't really ported to the platform.

**Comment by tschoonj** [comment]
Fair enough.

But I am not sure what to do with `tap_migrations.rb`. Formulas that move to core need to have their line in the file removed if present right? And those that move to gui need to have their line modified if present and added otherwise?

**Comment by DomT4** [comment]
> Formulas that move to core need to have their line in the file removed if present right?

Yup!

> And those that move to gui need to have their line modified if present and added otherwise?

And yup! :smile:

**Comment by tschoonj** [comment]
And if the formula deals with a scientific package, even if it contains a gui (and possibly a library with API), move it to science instead?

**Comment by tschoonj** [comment]
Also, I would like to stick to one PR per tap for the migration/cleanup: I am dealing with 4 repositories as it is and multiple PRs will probably create merge conflict issues with `tap_migrations.rb` in my core PR if I start submitting multiple PRs over this.

Just to keep things manageable for me :smile:

**Comment by mikemcquaid** [comment]
> Also, I would like to stick to one PR per tap for the migration/cleanup: I am dealing with 4 repositories as it is and multiple PRs will probably create merge conflict issues with tap_migrations.rb in my core PR if I start submitting multiple PRs over this.

Sure :+1: 

> And if the formula deals with a scientific package, even if it contains a gui (and possibly a library with API), move it to science instead?

Yup :+1:

**Comment by ablyler** [comment]
@Homebrew/maintainers looks like the mavericks bot machine might need rebooted. It keeps getting hung up and timing out on builds.

**Comment by DomT4** [comment]
The machine itself seems fine at the moment. Mavericks tends to build a little slower than the other machines, and there have been a significantly higher than normal number of long builds today. I think it's just those two things bumping into each other.

**Comment by tschoonj** [comment]
timeout kicked in after one hour :cry: 

could you give me more time?

Thanks!

**Comment by DomT4** [comment]
I'll bump the timeout later when the CI isn't backed up.

**Comment by UniqMartin** [comment]
Judging from some of the formulae that are being migrated and that only 4 of 37 made it in the first hour, I suspect this will require a ginormous build. I'm not sure how practical it is to want to group everything together. I'm just thinking about the very realistic possibility this could fail on the first try and will then require to pass the bots again.

**Comment by tschoonj** [comment]
It's actually only 11 formulas that need a build in this PR. Most commits here only modify `tap_migrations.rb`.
The big rebuild will be in Homebrew/homebrew-gui#2

From previous experience I know that large PRs could take a few attempts to build everything properly, but I am a patient man :smile: 
Splitting it up in separate PRs would make things much harder as I would get merge conflicts due to the extensive modifications to `tap_migrations.rb`

**Comment by UniqMartin** [comment]
@tschoonj Sorry, I drew my conclusion too quickly without properly looking at your changes. I agree now that the way you approached this makes sense.

File: Library/Homebrew/tap_migrations.rb [hunk 2]
```diff
@@ -138,7 +132,8 @@
   "lilypond" => "homebrew/tex",
   "lmutil" => "homebrew/binary",
   "magit" => "homebrew/emacs",
-  "meld" => "homebrew/x11",
+  "mat" => "homebrew/boneyard",
```
Comment (DomT4) [inline Library/Homebrew/tap_migrations.rb]: This line needs fixing, we moved `mat` back to the core. I presume this is a rebase gone slightly wrong or something.

**Comment by DomT4** [comment]
Oh, this also needs rebasing on top of the master apparently. If you do that it'll trigger another CI build automatically.

**Comment by tschoonj** [comment]
Ok, but I should wait until the build-time has been extended right?

**Comment by DomT4** [comment]
I pushed it upwards of three hours already for the Boost build. Feel free to jump on the back of that :)

**Comment by tschoonj** [comment]
Done!

This will happen quite a lot unfortunately as tap_migrations.rb seems to be altered almost every day :cry:

**Comment by mikemcquaid** [comment]
You may have better luck splitting this into multiple PRs, one per formula.

**Comment by mikemcquaid** [comment]
Or at least smaller groups.

**Comment by mikemcquaid** [comment]
but if this passes: :+1:

**Comment by DomT4** [comment]
@BrewTestBot test this please

**Comment by DomT4** [comment]
Sorry, I kicked you off the queue earlier to deal with a Boost follow-up that had active breakage and needed fixing before people had much opportunity to run `brew update` heh. You should be back on the queue shortly.

**Comment by tschoonj** [comment]
By the way, I would be grateful if you could not merge in any PRs that modify `tap_migrations.rb` until the this PR is merged in (and the three other ones connected to it).

This way I won't have to rebase and rebuild everything :smile:

**Comment by tschoonj** [comment]
This latest build was successful: no problems apart from audit failures.

**Comment by DomT4** [comment]
@BrewTestBot test this please

**Comment by DomT4** [comment]
I haven't forgotten this. It got wedged behind the CI fun of recent days. I'll take another run at the PR trio later today, presuming things stay quiet.

**Comment by DomT4** [comment]
@BrewTestBot test this please

**Comment by tschoonj** [comment]
Everything looks ok apart from audit failures

**Comment by DomT4** [comment]
Bottles have decided to vanish here so re-queued, for what will be the final time.

**Comment by DomT4** [comment]
Merged in a bunch o' commits. Thanks once more Tom!

**Comment by tschoonj** [comment]
You're most welcome

Thank you for your perseverance!

**Pull request closed without merging**